File: src/main/java/de/espend/idea/php/annotation/doctrine/reference/DoctrineAnnotationFieldProvider.java
package de.espend.idea.php.annotation.doctrine.reference;

import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiReference;
import com.jetbrains.php.lang.psi.elements.ParameterList;
import com.jetbrains.php.lang.psi.elements.PhpAttribute;
import com.jetbrains.php.lang.psi.elements.PhpClass;
import com.jetbrains.php.lang.psi.elements.StringLiteralExpression;
import de.espend.idea.php.annotation.extension.parameter.AnnotationPropertyParameter;
import de.espend.idea.php.annotation.extension.PhpAnnotationReferenceProvider;
import de.espend.idea.php.annotation.extension.parameter.PhpAnnotationReferenceProviderParameter;
import de.espend.idea.php.annotation.pattern.AnnotationPattern;
import de.espend.idea.php.annotation.doctrine.reference.references.DoctrinePhpClassFieldReference;
import de.espend.idea.php.annotation.util.PhpElementsUtil;
import de.espend.idea.php.annotation.util.PhpPsiAttributesUtil;
import org.jetbrains.annotations.Nullable;

/**
 * @author Daniel Espendiller <daniel@espendiller.net>
 */
public class DoctrineAnnotationFieldProvider implements PhpAnnotationReferenceProvider {

    @Nullable
    @Override
    public PsiReference[] getPropertyReferences(AnnotationPropertyParameter annotationPropertyParameter, PhpAnnotationReferenceProviderParameter referencesByElementParameter) {

        if(annotationPropertyParameter.getType() != AnnotationPropertyParameter.Type.PROPERTY_VALUE) {
            return null;
        }

        String propertyName = annotationPropertyParameter.getPropertyName();
        if(propertyName == null || !(propertyName.equals("mappedBy") || propertyName.equals("inversedBy"))) {
            return null;
        }

        PsiElement parent = annotationPropertyParameter.getElement().getParent();


        if (parent instanceof ParameterList) {
            PsiElement phpAttribute = parent.getParent();
            if (phpAttribute instanceof PhpAttribute) {
                String targetEntityValue = PhpPsiAttributesUtil.getAttributeValueByNameAsStringWithClassConstant((PhpAttribute) phpAttribute, "targetEntity");
                if (targetEntityValue != null) {
                    PhpClass phpClass = PhpElementsUtil.getClassInterface(annotationPropertyParameter.getProject(), targetEntityValue);
                    if(phpClass == null) {
                        return null;
                    }

                    return new PsiReference[] {
                        new DoctrinePhpClassFieldReference((StringLiteralExpression) annotationPropertyParameter.getElement(), phpClass)
                    };
                }
            }
        } else {
            StringLiteralExpression targetEntity = PhpElementsUtil.getChildrenOnPatternMatch(parent, AnnotationPattern.getPropertyIdentifierValue("targetEntity"));
            if(targetEntity == null) {
                return null;
            }

            PhpClass phpClass = PhpElementsUtil.getClassInsideAnnotation(targetEntity);
            if(phpClass == null) {
                return null;
            }

            return new PsiReference[] {
                new DoctrinePhpClassFieldReference((StringLiteralExpression) annotationPropertyParameter.getElement(), phpClass)
            };
        }

        return null;
    }
}

File: src/main/kotlin/de/espend/idea/php/annotation/doctrine/reference/DoctrineAnnotationFieldProvider.kt
package de.espend.idea.php.annotation.doctrine.reference

import com.intellij.psi.PsiReference
import com.jetbrains.php.lang.psi.elements.ParameterList
import com.jetbrains.php.lang.psi.elements.PhpAttribute
import com.jetbrains.php.lang.psi.elements.StringLiteralExpression
import de.espend.idea.php.annotation.doctrine.reference.references.DoctrinePhpClassFieldReference
import de.espend.idea.php.annotation.extension.PhpAnnotationReferenceProvider
import de.espend.idea.php.annotation.extension.parameter.AnnotationPropertyParameter
import de.espend.idea.php.annotation.extension.parameter.PhpAnnotationReferenceProviderParameter
import de.espend.idea.php.annotation.pattern.AnnotationPattern
import de.espend.idea.php.annotation.util.PhpElementsUtil
import de.espend.idea.php.annotation.util.PhpPsiAttributesUtil

/**
 * @author Daniel Espendiller <daniel@espendiller.net>
 */
class DoctrineAnnotationFieldProvider : PhpAnnotationReferenceProvider {
    override fun getPropertyReferences(
        annotationPropertyParameter: AnnotationPropertyParameter,
        referencesByElementParameter: PhpAnnotationReferenceProviderParameter?
    ): Array<PsiReference>? {
        if (annotationPropertyParameter.type != AnnotationPropertyParameter.Type.PROPERTY_VALUE) {
            return null
        }

        val propertyName = annotationPropertyParameter.propertyName
        if (propertyName != "mappedBy" && propertyName != "inversedBy") {
            return null
        }

        val parent = annotationPropertyParameter.element.parent

        if (parent is ParameterList) {
            val phpAttribute = parent.parent
            if (phpAttribute is PhpAttribute) {
                val targetEntityValue = PhpPsiAttributesUtil.getAttributeValueByNameAsStringWithClassConstant(
                    phpAttribute, "targetEntity"
                )
                if (targetEntityValue != null) {
                    val phpClass = PhpElementsUtil.getClassInterface(
                        annotationPropertyParameter.project,
                        targetEntityValue,
                    ) ?: return null

                    return arrayOf(
                        DoctrinePhpClassFieldReference(
                            annotationPropertyParameter.element as StringLiteralExpression,
                            phpClass
                        )
                    )
                }
            }
        } else {
            val targetEntity = PhpElementsUtil.getChildrenOnPatternMatch<StringLiteralExpression?>(
                parent,
                AnnotationPattern.getPropertyIdentifierValue("targetEntity")
            ) ?: return null

            val phpClass = PhpElementsUtil.getClassInsideAnnotation(targetEntity) ?: return null

            return arrayOf(
                DoctrinePhpClassFieldReference(
                    annotationPropertyParameter.element as StringLiteralExpression,
                    phpClass
                )
            )
        }

        return null
    }
}
